File: src/kv2kl/src/metno/kvalobs/kl2kv/DataHelper.java
/*
  Kvalobs - Free Quality Control Software for Meteorological Observations 

  $Id: DataHelper.java,v 1.1.2.5 2007/09/27 09:02:19 paule Exp $                                                       

  Copyright (C) 2007 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: kvalobs-dev@met.no

  This file is part of KVALOBS

  KVALOBS is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as 
  published by the Free Software Foundation; either version 2 
  of the License, or (at your option) any later version.
  
  KVALOBS is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License along 
  with KVALOBS; if not, write to the Free Software Foundation Inc., 
  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
package metno.kvalobs.kl2kv;

import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Timestamp;
import metno.kvalobs.kl.KlApp;
import metno.util.MiGMTTime;
import org.apache.log4j.Logger;

import sun.beans.editors.FloatEditor;
import metno.dbutil.DbConnection;
import metno.dbutil.DbConnectionMgr;


/**
 * This class is used to create a message that is to be sendt to the 
 * kldecoder in kvalobs. The data is read from the table KL2KVALOBS 
 * in an open database.
 * 
 * The format on the message to be sendt is.
 * <pre>
 	<paramkode1>[(<sensor1>,<level1>)][,<paramkode2>,...,<paramkodeN(<sensorN,levelN)>]
    YYYYMMDDHHMMSS,<original1i>,<original2i>,...,<originalNi>
    YYYYMMDDHHMMSS,<original1j>,<original2j>,...,<originalNj>
    ...
    YYYYMMDDHHMMSS,<original1n>,<original2n>,...,<originalNn>
 
 	Where the first line is a header that describes the data. The parameters
 	(paramkode) is the name of the parameter, not the paramid.
 	
 	The data in the datasection have the format:
 	original: value[(controlinfo,useinfo)] 
 	
 	Note that if controlinfo and useinfo is given, most of the values 
 	is set to default values before the qa controls, only a few values
 	is accepted and not changed. See the flaggdocument.
  </pre>
 * 
 *  
 * 
 * 
 * @author borgem
 *
 */
public class DataHelper{
	int MAX_COUNT;
		
	static Logger logger = Logger.getLogger(DataHelper.class);
	static final float MISSING_VAL=-32767.0f;
	String tablename="KL2KVALOBS";
	String typeid;
    String params;
    DataToKv dataToKv;
    DbConnection  con;
    int           msgCount;
    int           obsCount;
    static String[] ignoreList={"REG_DATO"};
    
    
    
    public String getTable(){
    	return tablename;
    }

    public String getTypeid(){
    	return typeid;
    }
    
    public boolean ignore(String param){
    	for(int i=0; i<ignoreList.length; i++)
    		if(ignoreList[i].equalsIgnoreCase(param))
    			return true;
    	
    	return false;
    }
    
    public boolean floatIsEq(float f1, float f2){
    	System.out.println("floatIsEq: "+f1 +", "+f2+" dif: "+Math.abs(f1-f2));
    	
    	if(Math.abs(f1-f2)<0.001)
    		return true;
    	else
    		return false;
    }
    
    public DataHelper(DbConnection con, DataToKv dataToKv,  
	                   String typeid){
    	this(con, dataToKv, typeid, null, 20);
    }
    
    public DataHelper(DbConnection con, DataToKv dataToKv,  
                       String typeid, String table){
    	this(con, dataToKv, typeid, table, 20);
    }
    
    /**
     * 
     * @param con A valid DbConnection.
     * @param dataToKv An instancs to a class that implements DataToKv.
     * @param typeid The typeid to the stations we shall send to kvalobs.
     * @param table The table we shall pull the data to kvalobs from. If 
     *         null is given the default table KL2KVALOBS is used.
     * @param max_count The data to be sendt to kvalobs is
     *        split up in max_count chuncks. 
     */
    public DataHelper(DbConnection con, DataToKv dataToKv,  
    		           String typeid, String table, int max_count){
    	params=null;
    	msgCount=0;
    	obsCount=0;
    	
    	if(table!=null)
    		this.tablename=table;
    	
    	if(max_count<1)
    		MAX_COUNT=1;
    	else
    		MAX_COUNT=max_count;
    	
    	this.typeid=typeid;
    	this.dataToKv=dataToKv;
    	this.con=con;
    }
    
    protected String createQuery(Station st){
    	String query=st.query();
    	
    	if(query.length()>0)
    		query=" AND "+query;
    	
   		return "Select * from "+ getTable()+
    				" where typeid="+getTypeid()+ query+
    				" order by stnr,dato";
    }

    /**
     * Before this method is called must createParams has been called 
     * at least one time. 
     * 
     * @return The header section to the message that is to be sendt
     * to kvalobs.
     */
    protected String getParams(){
    	return params;
    }
    
    
    /**
     * createParams creates a header for the message sendt to kvalobs.
     * The precipitation elements is added as the last element.
     * 
     * The creation of the header must match the data creation in
     * convertToKvDataAndSend. 
     * 
     * @param md Metadata
     * @return A string to use as a header in the message to kvalobs.
     * @throws SQLException
     */
    protected String createParams(ResultSetMetaData md) throws SQLException{
    	boolean first=true;
    	boolean hasRR24=false;

    	for(int i=4; i<=md.getColumnCount(); i++){
    		if(ignore(md.getColumnName(i)))
				continue;
				
    		if(md.getColumnName(i).equalsIgnoreCase("RR_24") ||
    		   md.getColumnName(i).equalsIgnoreCase("CONTROLINFO") ||
    		   md.getColumnName(i).equalsIgnoreCase("USEINFO")){
    			hasRR24=true;
    			continue;
    		}
    		
    		if(first){
    			params=md.getColumnName(i).toUpperCase();
    			first=false;
    		}else
    			params+=","+md.getColumnName(i).toUpperCase();
    	}
    	
    	if(hasRR24){
    		if(first)
    			params="RR_24";
    		else
    			params+=",RR_24";
    	}
    	
    	return params;
    }
    
    /**
     * This function creates a message to send to kvalobs. The data
     * is read from the resultset.
     * 
     * The creation of the data section must match the header that 
     * is created in createParams.
     * @param rs The dataset to send to kvalobs.
     * @return true on success.
     */
    protected boolean convertToKvDataAndSend(java.sql.ResultSet rs){
    	try{
    		ResultSetMetaData md=rs.getMetaData();
    		int nColumns=md.getColumnCount();
    		int stationid=-1;
    		int prevStationid=-1;
    		int count=0;
    		boolean hasData;
    		String tmp;
    		String data=createParams(md)+"\n";
    		String tmpData=null;
    		Timestamp tsobstime;
    		MiGMTTime obstime;
    		String RR24;
    		String cinfo_RR24;
    		String uinfo_RR24;
    		
    		while(rs.next()){
    			
    			stationid =  rs.getInt(1);

    			if((stationid!=prevStationid && prevStationid!=-1 && count>0) ||
    			    count>=MAX_COUNT){
    				System.out.println("DEBUG:"+prevStationid+"("+getTypeid()+") ["+
                                       data+"]");
    				if(!dataToKv.sendData(data, prevStationid, 
    				                     Integer.parseInt(getTypeid()))){
    					System.out.println("Cant send data to kvalobs!");
    					return false;
    				}
    				msgCount++;
    				data=getParams()+"\n";
    				count=0;
    			}

    			
    			prevStationid=stationid;
    			hasData=false;
    			
    			tsobstime=rs.getTimestamp(2);
    			
    			if(rs.wasNull()){
    				logger.error("No timestamp given (SQL NULL)!");
    				continue;
    			}
    			
    			obstime=new MiGMTTime(tsobstime);
    			tmpData=obstime.toString(MiGMTTime.FMT_COMPACT_TIMESTAMP_1);

    			hasData=false;
    			RR24="";
    			cinfo_RR24="";
    			uinfo_RR24="";
    			int i;
    			boolean missing=false;
    			
    			for (i = 4; i<=nColumns; i++) {
    				if(ignore(md.getColumnName(i)))
    					continue;

    				if(md.getColumnName(i).equalsIgnoreCase("RR_24")){
    					RR24=rs.getString(i);
    					
    					if(rs.wasNull())
    						RR24="";
    					else{
    						float f=Float.parseFloat(RR24);
    						
    						if(floatIsEq(f, MISSING_VAL)){
    							missing=true;
    						}
    					}
    				}else if(md.getColumnName(i).equalsIgnoreCase("CONTROLINFO")){
    					cinfo_RR24=rs.getString(i);
    					
    					if(rs.wasNull() || cinfo_RR24.length()!=16){
    						if(missing)
    							cinfo_RR24="0000003000000000";
    						else
    							cinfo_RR24="";
    					}else{
    						if(missing){
   								char buf[]=cinfo_RR24.toCharArray();
    							
   								if(buf.length!=16){
   									logger.warn("controlinfo (RR_24). Wrong size '"+buf.length+"' expected length 16.");
   									cinfo_RR24=null;
   								}else{
   									buf[6]='3';
   									cinfo_RR24=new String(buf);
   								}
    						}
    					}
    				}else if(md.getColumnName(i).equalsIgnoreCase("USEINFO")){
    					uinfo_RR24=rs.getString(i);
    					
    					if(rs.wasNull())
    						uinfo_RR24="";
    				}else{
    					tmp=rs.getString(i);
		        				
    					if(rs.wasNull()){
    						tmp="";
    					}else if(tmp.length()>0){
    						try {
								float f=Float.parseFloat(tmp);
			
								if(floatIsEq(f, MISSING_VAL))
										tmp=tmp+"(0000003000000000,)";
							} catch (NumberFormatException e) {
								logger.debug("NumberFormatException: '"+tmp+ "' Not a number (Float)");
							}
    					}
    					
    					tmpData+=","+tmp;

						if(tmp.length()>0)
							hasData=true;
    				}
    			}

                if(i>4){
                	tmpData+=",";
                	if(RR24.length()>0){
                		tmpData+=RR24+"("+cinfo_RR24+","+uinfo_RR24+")";
                		hasData=true;
                	}
   				}
    			
    			if(hasData){
    				data+=tmpData+"\n";
    				count++;
    				obsCount++;
    			}else{
        			System.out.println("NODATA: "+stationid+" ("+getTypeid()+") obstime: "+rs.getString(2));
        		}
    		}
	    
    		if(count>0){
    			System.out.println(stationid+"("+getTypeid()+") ["+
    							   data+"]");
    			if(!dataToKv.sendData(data, stationid, 
				   Integer.parseInt(getTypeid()))){
    				System.out.println("Cant send data to kvalobs!");
    				return false;
    			}
    			
    			msgCount++;
    		}
	    
    		return true;
    	}
    	catch(SQLException SQLe){                
    		System.out.println(SQLe);            
    		return false;
    	}
    }
    
    public boolean sendDataToKv(Station station){
    	System.out.println("Running for typeid: " + typeid);
    	System.out.println("table:       " + tablename);
    	System.out.println("Station(s): "+station);

    	logger.info("Running for typeid: " + typeid);
    	logger.info("table:       " + tablename);
    	logger.info("Station(s):  " + station);
    	
    	if(con==null){
    		System.out.println("No valid (null) connection to the db!");
     	    logger.error("No valid (null) connection to the db!");
     	    return false;
    	}
    	
    	String query=createQuery(station);
    	
    	System.out.println("Query:[" + query +"]");

    	if(query==null){
    	    System.out.println("Cant create query!!!");
    	    logger.error("Cant create query!!!");
    	    
    	    return false;
    	}

    	ResultSet rs;
    	
    	try{
    	    rs=con.execQuery(query);
    	}
    	catch(SQLException ex){
    	    System.out.println("Cant execute query:");
    	    System.out.println("[" +query+"]");
    	    System.out.println("Reason: "+ex);
    	    logger.error("Cant execute query:");
    	    logger.error("[" +query+"]");
    	    logger.error("Reason: "+ex);

    	    return false;
    	}
    	    
           
    	boolean res=convertToKvDataAndSend(rs);

    	return res;
    }
    
    public int getObsCount(){ return obsCount; }
    public int getMsgCount(){ return msgCount; }
}

File: src/tools/kv2kvDataInputd/DataHelper.cc
/*
  Kvalobs - Free Quality Control Software for Meteorological Observations 

  $Id$                                                       

  Copyright (C) 2007 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: kvalobs-dev@met.no

  This file is part of KVALOBS

  KVALOBS is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as 
  published by the Free Software Foundation; either version 2 
  of the License, or (at your option) any later version.
  
  KVALOBS is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License along 
  with KVALOBS; if not, write to the Free Software Foundation Inc., 
  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include <map>
#include <set>
#include <iostream>
#include <boost/algorithm/string.hpp>
#include <kvcpp/kvevents.h>
#include "DataHelper.h"

using namespace std;
using namespace miutil;
using namespace std;
using namespace kvservice;


Param::
Param(const Param &p )
   : paramid( p.paramid ), sensor( p.sensor ), level( p.level )
{
}

Param::
Param(const kvalobs::kvData &data )
   : paramid( data.paramID() ),
     sensor( data.sensor()-'0' ),
     level( data.level() )
{
}

Param::
Param( const kvalobs::kvTextData &textData )
   : paramid( textData.paramID() ),
     sensor( INT_MAX ),
     level( INT_MAX )
{
}

bool
Param::
isTextData()const
{
   return sensor == INT_MAX && level == INT_MAX;
}

Param
Param::
operator=(const Param &rhs )
{
   if( &rhs != this ) {
      paramid = rhs.paramid;
      sensor = rhs.sensor;
      level = rhs.level;
   }
   return *this;
}


bool
Param::
operator<(const Param &rhs )const
{

   if( paramid < rhs.paramid ||
       ( paramid == rhs.paramid && sensor < rhs.sensor) ||
       ( paramid == rhs.paramid && sensor == rhs.sensor && level < rhs.level )
     ){
      return true;
   }else {
      return false;
   }
}


bool
Param::
operator==(const Param &rhs )const
{
   if( paramid == rhs.paramid &&
       sensor == rhs.sensor &&
       level == rhs.level ){
      return true;
   }else {
      return false;
   }
}


std::ostream&
operator<<( std::ostream &out, const Param &param )
{
   out << "[" << param.paramid << "," << param.sensor << ","
       << param.level << "]";
   return out;
}




ObsData::
ObsData()
   :stationid( INT_MAX ), type( INT_MAX )
{
}

ObsData::
ObsData( int stationid, int typeid_ )
   : stationid( stationid ), type( type )
{
}

bool
ObsData::
add( const kvalobs::kvData &kvData )
{
   if( stationid == INT_MAX ) {
      stationid = kvData.stationID();
      type = kvData.typeID();
   } else if( stationid != kvData.stationID() ||
              type != kvData.typeID() ) {
      return false;
   }

   if( kvData.cfailed().find("hqc") != string::npos ) {
      hqcTimes.insert( kvData.obstime() );
      return true;
   }

   data[kvData.obstime()][Param(kvData)]=kvData.original();
   header.insert( Param( kvData ) );
   obsTimes.insert( kvData.obstime() );

   return true;
}

bool
ObsData::
add( const kvalobs::kvTextData &kvData )
{
   if( stationid == INT_MAX ) {
      stationid = kvData.stationID();
      type = kvData.typeID();
   } else if( stationid != kvData.stationID() ||
              type != kvData.typeID() ) {
      return false;
   }

   if( hqcTimes.find( kvData.obstime() ) != hqcTimes.end() ) {
      return true;
   }

   textData[kvData.obstime()][Param(kvData)] = kvData.original();
   header.insert( Param( kvData ) );
   obsTimes.insert( kvData.obstime() );

   return true;
}

std::string
ObsData::
createHeader( ParamDefsPtr paramdefs )
{
   ostringstream ost;
   bool first=true;

   for( Params::iterator it=header.begin(); it != header.end(); ++it ){
      ParamDefs::iterator hit = paramdefs->find( it->paramid );

      //Remove element that we do NOT recognize.
      if( hit == paramdefs->end() ) {
         header.erase( it );
         continue;
      }

      if( first ) {
         first = false;
      } else {
         ost << ",";
      }

      ost << hit->second;

      if( ! it->isTextData() && (it->sensor != 0 || it->level != 0) ) {
         ost << "(" << it->sensor << "," << it->level << ")";
      }
   }

   //No params?
   if( first )
      return "";

   ost << endl;
   return ost.str();
}

std::string
ObsData::
getData( const Param &param, const miutil::miTime &obstime )const
{
   if( param.isTextData() ) {
      TextDataList::const_iterator obstimeIt = textData.find( obstime );

      if( obstimeIt == textData.end() )
         return "";

      std::map<Param, std::string>::const_iterator paramIt;
      paramIt = obstimeIt->second.find( param );

      if( paramIt == obstimeIt->second.end() )
         return "";

      string data=paramIt->second;
      string tmp = boost::replace_all_copy( data, ",", "\\,");

      if( tmp != data )
         return "\"" + tmp +"\"";
      else
         return data;
   } else {
      DataList::const_iterator obstimeIt = data.find( obstime );

      if( obstimeIt == data.end() )
         return "";

      std::map<Param, float>::const_iterator paramIt;
      paramIt = obstimeIt->second.find( param );

      if( paramIt == obstimeIt->second.end() )
         return "";

      ostringstream ost;
      ost << paramIt->second;
      return ost.str();
   }
}

void
ObsData::
klData( std::string &data, std::string &decoder,
        ParamDefsPtr paramdefs )
{
   string headerString = createHeader( paramdefs );

   data.clear();
   decoder.clear();

   if( headerString.empty() ) {
      return;
   }

   ostringstream ost;
   int count=0;
   int nParams;
   ost << headerString;

   for( Times::iterator it=obsTimes.begin(); it != obsTimes.end(); ++it ) {
      //We do not generate a message if the hqc has been involved.
      if( hqcTimes.find( *it ) != hqcTimes.end() )
         continue;

      nParams = 0;
      for( Params::iterator pit = header.begin(); pit != header.end(); ++pit ) {
         if( nParams == 0 )
            ost << *it;

         ost << ",";

         ++nParams;
         ost << getData( *pit, *it );
      }

      if( nParams > 0 ) {
         ost << endl;
         ++count;
      }
   }

   if( count > 0 ) {
      data = ost.str();
      ost.str("");
      ost << "kldata/nationalnr=" << stationid <<"/type=" << type;
      decoder = ost.str();
   }
}

bool
ObsData::
operator<(const ObsData &rhs )const
{
   if( stationid < rhs.stationid && type < rhs.type )
      return true;
   else
      return false;
}

std::ostream&
operator<<( std::ostream &out, const ObsData &od)
{
   out << "[(";
   for( ObsData::Params::const_iterator hit=od.header.begin();
        hit != od.header.end(); ++hit) {
      if( hit != od.header.begin() )
         out << ",";
      out << *hit;
   }
   out<< ")" << endl;
   out << "(";
   for( ObsData::DataList::const_iterator dit=od.data.begin();
        dit != od.data.end(); ++dit ) {
      out << dit->first;
      for(std::map<Param, float>::const_iterator pit=dit->second.begin();
            pit != dit->second.end(); ++pit ) {
         out << ","<<pit->first << ":"<<pit->second;
      }
      out << endl;
   }
   out << ")"  << endl;
   out << "(";
   for( ObsData::TextDataList::const_iterator dit=od.textData.begin();
         dit != od.textData.end(); ++dit ) {
      out << dit->first;
      for(std::map<Param, string>::const_iterator pit=dit->second.begin();
            pit != dit->second.end(); ++pit ) {
         out << ","<<pit->first << ":"<<pit->second;
      }
      out << endl;
   }
   out << ")"  << endl;



}

DataHelper::
DataHelper()
   : itNext( dataList.end() )
{
}

void
DataHelper::
addData( KvObsDataListPtr obs )
{
   for( KvObsDataList::iterator it = obs->begin();
        it != obs->end(); ++it ) {
      for( KvObsData::kvDataList::iterator dit=it->dataList().begin();
           dit != it->dataList().end(); ++dit ) {
         if( dit->typeID() < 0 || dit->original() <= -32766 )
            continue;

         obsData[dit->stationID()][dit->typeID()].add( *dit );
      }

      for( KvObsData::kvTextDataList::iterator dit=it->textDataList().begin();
           dit != it->textDataList().end(); ++dit ) {
         obsData[dit->stationID()][dit->typeID()].add( *dit );
      }
   }
   initNext();
}


void
DataHelper::
initNext()
{
   //Index the obsData to make it easy to get the encoded data

   dataList.clear();

   for( ObsDataList::iterator it=obsData.begin();
         it != obsData.end(); ++it ) {
      for( map<int, ObsData>::iterator itt = it->second.begin();
            itt != it->second.end(); ++itt )
         dataList.push_back( &itt->second );
   }

   itNext = dataList.begin();
}



bool
DataHelper::
nextData( string &data, string &decoder, ParamDefsPtr paramdefs  )
{
   while( itNext != dataList.end() ) {
      (*itNext)->klData( data, decoder, paramdefs );
      ++itNext;

      if( ! data.empty() )
         return true;
      cerr << "DataHelper::nextData: empty data." << endl;
   }

   return false;
}

std::ostream&
operator<<( std::ostream &out, const DataHelper &dh )
{
   out << "------ [BEGIN DataHelper] ---------------" << endl;
   for( ObsDataList::const_iterator sit = dh.obsData.begin();
        sit != dh.obsData.end(); ++sit ) {
      for( map<int, ObsData>::const_iterator tit=sit->second.begin();
           tit != sit->second.end(); ++tit ) {
         out << tit->second << endl;
         out << "-------------------------------------" << endl;
      }

   }
   out << "------ [END DataHelper] ---------------" << endl;
}
